Add support for delete animation in LayoutAnimation on iOS

janicduplessis opened the pull request:
This adds support for delete view animations in LayoutAnimation for iOS. It supports the same properties as the create animation (alpha, scale).

This allows making simple animations when removing a view which is normally hard to do in React since we need to not remove the view node immediately.

**Test plan**
Tested add/removing views in the UIExample explorer with and without setting a LayoutAnimation. Also tested that the completion callback still works properly. Tested that user interation during the animation is properly disabled.

![layout-anim2](https://cloud.githubusercontent.com/assets/2677334/14595471/86fb1654-050d-11e6-8b38-fe45cc2dcd71.gif)

I also plan to work on improving the doc for LayoutAnimation as well as making this PR for android too.

janicduplessis commented on Libraries/LayoutAnimation/LayoutAnimation.js:
```diff
@@ -74,6 +74,12 @@ function configureNext(config: Config, onAnimationDidEnd?: Function) {
   UIManager.configureNextLayoutAnimation(
     config, onAnimationDidEnd || function() {}, function() { /* unused */ }
   );
+  // Clear the layout animation configuration after the current frame.
```
I moved reseting the layout animation configuration to JS since it needs to be reset at the end of the JS frame and not only in the `_layoutAndMount` function. Not sure if there is a hook in native that gets called at the end of the JS frame it might be better there.

janicduplessis commented on Libraries/LayoutAnimation/LayoutAnimation.js:
```diff
@@ -74,6 +74,12 @@ function configureNext(config: Config, onAnimationDidEnd?: Function) {
+  });
 }
 
 function create(duration: number, type, creationProp): Config {
```
Should we support setting a different prop for create and delete or this seems fine?

mkonicek commented:
Wow!

@sahrens, @vjeux What do you think? This seems pretty impactful.

vjeux commented:
Looks good to me, but I'll let @sahrens review it

janicduplessis commented:
@vjeux @sahrens I found an issue with resetting the layout anination when doing more testing. Sometimes it gets reset too early with requestAnimationFrame. If I do something like setTimeout 100 then it works perfectly but that's not a valid solution.

It still works most of the time so maybe it's an issue with the order the commands get dispatched through the bridge? Any idea on a way to make sure the layout animation gets reset after all the view operations are done?

I'll investigate this more this weekend.

sahrens commented:
This is awesome! Do you have a video of it working?

Hmm, if the layout takes more than one frame then I believe it's possible for requestAnimationFrame to fire before the layout completes. Can you use onLayout instead an verify that the layout has finished updating?

sahrens commented on React/Modules/RCTUIManager.m:
```diff
@@ -729,9 +729,44 @@ - (void)_amendPendingUIBlocksWithStylePropagationUpdateForShadowView:(RCTShadowV
+          RCTLogError(@"Unsupported layout animation createConfig property %@",
+                      deleteAnimation.property);
+        }
+      } withCompletionBlock:completion];
```
I'm worried about weird edge-cases that could happen if the user interacts with the app while this deletion animation is running. Perhaps we should make the view immediately non-interactable? Any other safeguards we could use? How much did you bang on the prototype?

sahrens commented:
cc @nicklockwood since he owns the layout animation implementation these days - any other concerns, Nick?

sahrens commented:
Requesting changes for you to 

* Beef up the test plan (maybe include an rnplay build for other people bang on?)
* Fix the requestAnimationFrame stuff...definitely don't want any race conditions in there. Might need some special native logic to get it working right if you can't get onLayout to work 100%.

janicduplessis commented:
Thanks @sahrens for the review, I did some more debugging and found a way to clear the layout animation properly.

Good point about edge cases I did some more testing by making the transition super slow and added some Touchable and native views inside the view that gets deleted with an animation. It doesn't cause any issue, for touchable (and probably any js driven view) it just stop responding to interactions, for native views (Switch for example) it is still possible to interact with it but no nothing gets sent to JS.

Do we want to set userInterationsEnabled = false here to have the same behaviour for all the views when they are getting removed with an animation?

I don't think I can put an example on RNPlay since it needs native code from this PR to work. I did some testing using the example I added to UIExplorer and made the animation super slow.
https://gist.github.com/janicduplessis/8e0dc3a711113006a55ae7f6f04dd45a

janicduplessis commented:
Here's a video of it.

![layout-anim](https://cloud.githubusercontent.com/assets/2677334/14595409/ea84e46c-050c-11e6-8990-4a12a236c754.gif)

I added a cross fade example too but it could be improved because the view that gets removed should not fade out, it should just stay untill the animation is over, I was thinking adding a delete animation the just makes the view stays untill the LayoutAnimation is over and add a CrossFade preset. However, it should probably be in another PR.

mkonicek commented:
@sahrens Thanks a lot for the review! Is this good to go now?

sahrens commented on Examples/UIExplorer/LayoutAnimationExample.js:
```diff
@@ -0,0 +1,169 @@
+          {
+            this.state.toggled ?
+            <GreenSquare /> :
+            <BlueSquare />
```
This is pretty sick how easy this is.

sahrens commented on React/Modules/RCTUIManager.m:
```diff
@@ -848,8 +884,8 @@ - (void)_manageChildren:(NSNumber *)containerReactTag
-  [self _removeChildren:permanentlyRemovedChildren fromContainer:container];
-  [self _removeChildren:temporarilyRemovedChildren fromContainer:container];
+  [self _removeChildren:permanentlyRemovedChildren fromContainer:container permanent: true];
+  [self _removeChildren:temporarilyRemovedChildren fromContainer:container permanent: false];
```
nit: no space after colons.

sahrens commented on Examples/UIExplorer/LayoutAnimationExample.js:
```diff
@@ -0,0 +1,169 @@
+
+  _onPressToggle() {
+    LayoutAnimation.easeInEaseOut();
+    this.setState({toggled: !this.state.toggled});
```
nit: use state function:

  this.setState((state) => ({toggled: !state.toggled}));

janicduplessis replied:
Wow, I didn't know you could do that, looks a lot cleaner :)

sahrens replied:
It's also more correct because another state update might be queued but not yet reflected in this.state, using the function makes sure you are operating on the latest state.

sahrens commented on Examples/UIExplorer/LayoutAnimationExample.js:
```diff
@@ -0,0 +1,169 @@
+
+  _onPressAddView() {
+    LayoutAnimation.easeInEaseOut();
+    this.setState({views: [...this.state.views, {}]});
```
These should also use state mutation functions.

sahrens commented on Examples/UIExplorer/LayoutAnimationExample.js:
```diff
@@ -0,0 +1,169 @@
+  },
+
+  _onPressRemoveView() {
+    LayoutAnimation.easeInEaseOut();
```
Another approach would be to add `LayoutAnimat.easeInEaseOut()` to `componentWillUpdate()`.

sahrens commented:
This looks awesome, just some nits.

I'm not sure I understand what you're saying about the cross-fade - what's wrong with the one deleted fading out?

I do agree in general that there should be more configuration options, though. Looking forward to more PR's!

sahrens commented:
Also, can you put the video in the test plan?

sahrens commented:
Have you had a chance to look at the Android implementation yet? Any estimate on how long that might take?

janicduplessis commented:
It's hard to see in the video but in the middle of the animation the view becomes a bit white (or whatever the background color is under the cross fading views). It happens because the old view alpha is at 0.5 and the new view alpha is at 0.5 too so the background color behind those 2 views become visible. To avoid this we can just animate the alpha of the new view and just keep the old one around untill the animation is over.

As for the Android implementation I haven't looked at it yet so I have no idea if it's going to be simple or not.

sahrens commented:
Hmm, that makes sense for views that perfectly overlap, but not sure how often that would be the case. Having that crossfade option would be handy there though.


On Apr 22, 2016, at 9:01 AM, Janic Duplessis <notifications@github.com<mailto:notifications@github.com>> wrote:


It's hard to see in the video but in the middle of the animation the view becomes a bit white (or whatever the background color is under the cross fading views). It happens because the old view alpha is at 0.5 and the new view alpha is at 0.5 too so the background color behind those 2 views become visible. To avoid this we can just animate the alpha of the new view and just keep the old one around untill the animation is over.

As for the Android implementation I haven't looked at it yet so I have no idea if it's going to be simple or not.

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub<https://github.com/facebook/react-native/pull/6779#issuecomment-213490428>

janicduplessis commented:
Fixed the nits and also added userInterationEnabled = NO since JS doesn't receive the events anyway it could be confusing if you click on a switch and it moves but JS never received anything.

sahrens commented:
Awesome, thanks!

@facebook-github-bot shipit

sahrens commented:
Looked like tests failed - can you rebase and if there are still errors, fix them?

janicduplessis commented:
Travis tests seems fine now lets try it again.

@facebook-github-bot shipit

dwilt commented:
Did this get merged? I'm on `0.59.3` and still not seeing `delete` animations but am seeing `create` animations while using `LayoutAnimation.easeInEaseOut()`
